controllers/secretdefinition_controller.go

```diff
@@ -0,0 +1,242 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	secretsmanagerv1alpha1 "github.com/tuenti/secrets-manager/api/v1alpha1"
+	"github.com/tuenti/secrets-manager/backend"
+)
+
+const (
+	// https://golang.org/pkg/time/#pkg-constants
+	timestampFormat = "2006-01-02T15.04.05Z"
+	finalizerName   = "secret.finalizer.secrets-manager.tuenti.io"
+)
+
+// SecretDefinitionReconciler reconciles a SecretDefinition object
+type SecretDefinitionReconciler struct {
+	client.Client
+	Backend              backend.Client
+	Log                  logr.Logger
+	Ctx                  context.Context
+	APIReader            client.Reader
+	ReconciliationPeriod time.Duration
+}
+
+// Helper functions to check and remove string from a slice of strings.
+func containsString(slice []string, s string) bool {
+	for _, item := range slice {
+		if item == s {
+			return true
+		}
+	}
+	return false
+}
+
+func removeString(slice []string, s string) (result []string) {
+	for _, item := range slice {
+		if item == s {
+			continue
+		}
+		result = append(result, item)
+	}
+	return
+}
+
+// Ignore not found errors
+func ignoreNotFoundError(err error) error {
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+// getDesiredState reads the content from the Datasource for later comparison
+func (r *SecretDefinitionReconciler) getDesiredState(keysMap map[string]secretsmanagerv1alpha1.DataSource) (map[string][]byte, error) {
+	desiredState := make(map[string][]byte)
+	var err error
+	for k, v := range keysMap {
+		bSecret, err := r.Backend.ReadSecret(v.Path, v.Key)
```

Comment:
(feeling fancy) this comment is extremely _tuenti style_ , just saying this in case you feel **t-** fancy today ;)

1) from the v you get the `bSecret`
2) from the v you get the `decoder`
3) with `bSecret` & `decoder` you get the response

go is super fancy for executing two functions in parallel and then receive both values, seems that 1,2 can be done with the `go` keyword

_abstract_

```go
    out1 := make(chan XXX)
    out2 := make(chan XXX)

    go func() {
        out1 <- f1()
    }()

    go func() {
        out2 <- f2()
    }()

    foo(<-out1, <-out2)
```

Reply:
I will disgree on this, seems like way cleaner just reading from the kv.